documents/model.go

```diff
@@ -754,3 +758,153 @@ func isNFTInRole(role *coredocumentpb.Role, registry common.Address, tokenID []b
 
 	return false
 }
+
+// assembleAccessToken assembles a Read Access Token from the payload received
+func assembleAccessToken(dm *CoreDocumentModel, id identity.IDConfig, payload documentpb.AccessTokenParams) (*coredocumentpb.AccessToken, error) {
+	tokenIdentifier := utils.RandomSlice(32)
+	granterID := id.ID[:]
+	roleID := byte(len(dm.Document.Roles))
+	grantee, err := hexutil.Decode(payload.GetGrantee())
+	if err != nil {
+		return nil, err
+	}
+	docID, err := hexutil.Decode(payload.GetDocumentIdentifier())
+	if err != nil {
+		return nil, err
+	}
+	// assemble access token message to be signed
+	tm := append(tokenIdentifier, granterID...)
+	tm = append(tm, grantee...)
+	tm = append(tm, roleID)
+	tm = append(tm, docID...)
+
+	// fetch key pair from identity
+	privateKey := id.Keys[identity.KeyPurposeEthMsgAuth].PrivateKey
+	pubKey := id.Keys[identity.KeyPurposeEthMsgAuth].PublicKey
+
+	// sign the assembled access token message, return signature
+	sig, err := crypto.SignMessage(privateKey, tm, "CurveSecp256K1", true)
+	if err != nil {
+		return nil, err
+	}
+	// assemble the access token, appending the signature and public keys
+	at := new(coredocumentpb.AccessToken)
+	at.Identifier = tokenIdentifier
+	at.Granter = granterID
+	at.Grantee = grantee
+	at.RoleIdentifier = []byte{roleID}
+	at.DocumentIdentifier = docID
+	at.Signature = sig
+	at.Key = pubKey
+
+	return at, nil
+}
+
+// AddAccessTokenToReadRules adds the AccessToken(s) to the read rules of the document
+func (m *CoreDocumentModel) AddAccessTokenToReadRules(id identity.IDConfig, payload documentpb.AccessTokenParams) error {
+	at, err := assembleAccessToken(m, id, payload)
+	if err != nil {
+		return errors.New("failed to construct AT: %v", err)
+	}
+
+	role := new(coredocumentpb.Role)
```

Comment:
Role should be created first and then use the created role label in the access token